From f4b2d84b5ab4820d2a66b388c2c17c2a755af010 Mon Sep 17 00:00:00 2001
From: Sriram R <quic_srirrama@quicinc.com>
Date: Sat, 5 Nov 2022 13:14:14 +0530
Subject: [PATCH] ath12k: perform Tx address translation

When the link id is mentioned in the tx control info,
use it to convert the RA/TA to the corresponding
link sta and link vif address before transmission.

For eth data frames, always enqueue the frame on the
primary link/umac.

TODO: nwifi/raw data frame handling

Signed-off-by: Sriram R <quic_srirrama@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 64 +++++++++++++++++++++++++++
 1 file changed, 64 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 0394e4c..e643236 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7698,6 +7698,53 @@ static int ath12k_mac_mgmt_tx(struct ath12k *ar, struct sk_buff *skb,
 	return 0;
 }
 
+static u8 ath12k_mac_get_tx_link(struct ieee80211_sta *sta, struct ieee80211_vif *vif,
+				 u8 link, struct sk_buff *skb, u32 info_flags)
+{
+	struct ath12k_sta *ahsta;
+	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
+	struct ieee80211_link_sta *link_sta;
+	struct ieee80211_bss_conf *bss_conf;
+	struct ieee80211_hdr *hdr = (struct ieee80211_hdr *)skb->data;
+
+	/* Use the link id passed or the default vif link */
+	if (!sta) {
+		if (link != IEEE80211_LINK_UNSPECIFIED)
+			return link;
+
+		return ahvif->deflink.link_id;
+	}
+
+	ahsta = (struct ath12k_sta *)sta->drv_priv;
+
+	/* use default */
+	if (!sta->mlo)
+		return ahsta->deflink.link_id;
+
+	/* enqueue eth data frames on assoc link */
+	if (info_flags & IEEE80211_TX_CTL_HW_80211_ENCAP)
+		return ahsta->assoc_link_id;
+
+	/* 802.11 frame cases */
+	if (link == IEEE80211_LINK_UNSPECIFIED)
+		link = ahsta->deflink.link_id;
+
+	/* Perform address conversion for ML STA Tx */
+	bss_conf = rcu_dereference(vif->link_conf[link]);
+	link_sta = rcu_dereference(sta->link[link]);
+	if (bss_conf && link_sta) {
+		ether_addr_copy(hdr->addr1, link_sta->addr);
+		ether_addr_copy(hdr->addr2, bss_conf->addr);
+	} else {
+		WARN_ON(1);
+		link = IEEE80211_MLD_MAX_NUM_LINKS;
+	}
+
+	/* todo nwifi/raw data frames */
+	return link;
+
+}
+
 static void ath12k_mac_op_tx(struct ieee80211_hw *hw,
 			     struct ieee80211_tx_control *control,
 			     struct sk_buff *skb)
@@ -7720,6 +7767,7 @@ static void ath12k_mac_op_tx(struct ieee80211_hw *hw,
 	bool is_prb_rsp;
 	u16 frm_type = 0;
 	int ret;
+ 	u8 link = u32_get_bits(info->control.flags, IEEE80211_TX_CTRL_MLO_LINK);
 
 	if (ahvif->vdev_type == WMI_VDEV_TYPE_MONITOR) {
 		ieee80211_free_txskb(ah->hw, skb);
@@ -7734,6 +7782,22 @@ static void ath12k_mac_op_tx(struct ieee80211_hw *hw,
 	memset(skb_cb, 0, sizeof(*skb_cb));
 	skb_cb->vif = vif;
 
+	/* handle only for MLO case, use deflink for non MLO case */
+	if (vif->valid_links) {
+		link = ath12k_mac_get_tx_link(sta, vif, link, skb, info_flags);
+
+		if (WARN_ON(link >= IEEE80211_MLD_MAX_NUM_LINKS)) {
+			ieee80211_free_txskb(ah->hw, skb);
+			return;
+		}
+		if (WARN_ON(!ahvif->link[link])) {
+			ieee80211_free_txskb(ah->hw, skb);
+			return;
+		}
+		arvif = ahvif->link[link];
+		ar = arvif->ar;
+	}
+
 	if (key) {
 		skb_cb->cipher = key->cipher;
 		skb_cb->flags |= ATH12K_SKB_CIPHER_SET;
-- 
2.17.1

